Allow 32-bit PV drivers to work on 64-bit Xen.

Not used any more, as the underlying bugs which caused this param to be
created have been fixed properly upstream.

It served two purposes.  First to correctly latch the 32bit shinfo, and to
force a resychronisation of the domain wallclock, as qemu has likely just
stepped it on resume.

"Legacy" windows drivers unconditionally make this hypercall and ignore the
return value.

The XS6.1 and XS6.2 non-legacy drivers still make this hypercall, but
will BSOD with 0x0000DEAD if the call fails.

diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 9233765..696381f 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5629,6 +5629,7 @@ static int hvm_allow_set_param(struct domain *d,
     {
     /* The following parameters can be set by the guest. */
     case HVM_PARAM_CALLBACK_IRQ:
+    case HVM_PARAM_32BIT:
     case HVM_PARAM_VM86_TSS:
     case HVM_PARAM_ACPI_IOPORTS_LOCATION:
     case HVM_PARAM_VM_GENERATION_ID_ADDR:
@@ -5839,6 +5840,7 @@ static int hvm_allow_get_param(struct domain *d,
     {
     /* The following parameters can be read by the guest. */
     case HVM_PARAM_CALLBACK_IRQ:
+    case HVM_PARAM_32BIT:
     case HVM_PARAM_VM86_TSS:
     case HVM_PARAM_ACPI_IOPORTS_LOCATION:
     case HVM_PARAM_VM_GENERATION_ID_ADDR:
diff --git a/xen/include/public/hvm/params.h b/xen/include/public/hvm/params.h
index 7c73089..e940e3d 100644
--- a/xen/include/public/hvm/params.h
+++ b/xen/include/public/hvm/params.h
@@ -56,6 +56,8 @@
 
 #if defined(__i386__) || defined(__x86_64__)
 
+#define HVM_PARAM_32BIT        8
+
 /*
  * Viridian enlightenments
  *
